lib/buildConnectorsJson/generateSchemaFromJs.js

```diff
@@ -5,129 +5,106 @@
 var prettyTitle  = require('./prettyTitle');
 
-
-module.exports = function (config) {
-	return generateSchema(config, true);
+//Standard supported keys
+var supportedKeys = [
+	'type',
+	'title',
+	'enum',
+	'oneOf',
+	'name',
+	'default',
+	'description',
+	'lookup',
+	'format'
+];
+
+module.exports = function (rawSchema) {
+	var schema = generateSchema(rawSchema, true);
+	//Since $schema is only required at root level, assign it here.
+	schema['$schema'] = "http://json-schema.org/draft-04/schema#";
+	 // Standard restriction: top level schema never allows for additional properties
+	schema['additionalProperties'] = false;
+	return schema;
 };
 
-// Separate function down here so it can recursively call itself
-function generateSchema(config, topLevel) {
+//Generates the whole schema for a given  (but not necessarily the schema of a object value)
+function generateSchema(rawSchema, topLevel) {
 
 	var schema = {
-    type: "object",
-    properties: {}
+	    type: "object",
+	    properties: {},
+		required: []
 	};
 
-	if (topLevel === true) {
-		schema.$schema = "http://json-schema.org/draft-04/schema#";
-	}
-
-
-	var inputKeys = _.keys(config);
+	//Advanced only need for top level
+	if (topLevel) schema.advanced = [];
 
-	// Set variables as required
-	schema.required = _.filter(inputKeys, function (key) {
-		var messageInput = config[key];
-		return (_.isObject(messageInput) && messageInput.required === true);
+	//Set required and advanced variables respectively
+	_.forEach(rawSchema, function (valObj, key) {
+		if (_.isObject(valObj)) {
+			if (valObj.required === true) schema.required.push(key);
+			if (topLevel && valObj.advanced === true) schema.advanced.push(key);
+		}
 	});
 
-	// Set variables as advanced
-	schema.advanced = _.filter(inputKeys, function (key) {
-		var messageInput = config[key];
-		return (_.isObject(messageInput) && messageInput.advanced === true);
+	// Add the variables in
+	_.each(rawSchema, function (rawObject, key) {
+		schema.properties[key] = generateSchemaObject(rawObject, key);
 	});
 
-	// Standard restriction: top level schema never allows for additional properties
-	schema.additionalProperties = false;
-
-	// Add the variables in
-	_.each(config, function (param, key) {
-		var newParam = {};
+	return schema;
 
-		// Recursively generate the schema for objects
-		if (param.type === 'object') {
-			newParam = generateSchema(param.properties || {});
-		}
+}
 
-		// Add in all the fluff
-		_.extend(newParam, _.pick(param, [
-			'type',
-			'name',
-			'title',
-			'description',
-			'default',
-			'enum',
-			'additionalProperties',
-			'additionalItems',
-			'format',
-			'timeout_millis',
-      'lookup'
-		]));
-
-		// Un camelcase the default jsonpath
-		if (param.defaultJsonPath) {
-			newParam.default_jsonpath = param.defaultJsonPath;
-		}
+//The generation of a schema object is seperate, so that selective recusion can be applied
+function generateSchemaObject(rawObject, title) {
```

Comment:
Since the function is part of schema generation, I think it should remain part of the main generator file.